### DISCUSSION TITLE
Setup Assistance

### DISCUSSION THREAD

**Discussion by duckhead** [wgu-opensource/osmt#451] [Asking for Help] [answered]
So, I've been asked to look into a few OSMT issues by a very pretty lady sitting next to me. I've got lots of Java, but am a Docker noob. So, I'm just trying to be able to get the environment built and running, so I can poke around at it.

I've got a few questions (and, I'm happy to update the readme.md with the answers, I've done a lot of doc.) 

First, a note, Maven 3.8.6 no longer exists where you attempt to pull it from (I don't know why...) I had to update the script to 3.8.8. I've got that fixed in my local copy.

Secondly, I have attempted to get an OKTA account. And I think I was successful. But, I don't know how to get the information that you want in your file. I'd love to get a good explanation soup to nuts and I'll update the doc.

When I run docker-compose on dev-stack.yml, it looks like everything starts, but  I see that MySQL is angry that I didn't supply the MYSQL* environment variables. I need to fix that, but I'm not sure what I need to do. I see some references to that in the osmt_cli.sh script, but there it gets default values, and apparently it wasn't there at  the mysql container startup.

Finally,  this is kind of the end of the docs. I'm not sure what is next. What are the base and build containers?  The base container is just a JDK, wget, and curl? So, I use that as my dev environment? How do I do that with the sources not being in there? The build container is just Maven? How am I to use that, just to build?  Again, I'm happy to update the docs.

**Comment by duckhead** [comment]
OK, I've been hacking on this some more, and I think all of the above issues are resolved. 

It took me a bit to follow through the Okta documentation, and there was at least one choice I was asked to make that wasn't in the doc. I either did or did not guess correctly (I'll explain that later.)

I was able to build the solution, and actually run it using osmt_cli.sh. It starts up, and I can connect to it on localhost:8080 . That all is fine. From there, I pull up the webpage, and got the WGU page, with the spinner in the middle. But, it kept spinning. Eventually, I was taken to a Okta webpage complaining that I was apparently trying to authenticate too many times. So, I'm probably DOSing them.

I'd love to look at the logs, but I don't know where to find them. Does anyone have a pointer for me?

**Reply by JohnKallies** [reply]
Hello @duckhead. I'm glad you got some progress using the CLI tool. For a local dev instance, it's our intention that you just need Java/Maven and Docker (docker-compose provides the backing MySQL / ElasticSearch / Redis instances). We provide a fixed dataset in the GitHub repo for satisfying API test cases, and if you want to use the CLI to load this, you'll also need the MySQL client on your path as well. Please note that MySQL changes need propagated to ElasticSearch. Reindexing is done through a Spring command line runner `ReindexCommand`, and provided by `/.osmt_cli.sh -r`

The Spring logs should be in whatever terminal session you are running `./osmt_cli.sh -s`. If you want to set a deeper logging level, you can `export DEBUG=1` in that shell, or set this in your application-dev.properties (there are also some Spring config values that you can uncomment in that props file).

Hoping to bypass Okta for local dev and CI build purposes, we experimented with some kind of "no auth" config, or using a local mock OAuth2/OIDC that always returns a preconfigured JWT, but we were not yet successful with that.

I'd like your feedback on the gaps in the docs. It was hard to get them to say the right amount, and we may have moved beyond them in some areas not seeing the gaps. A fresh set of eyes might help there.We invite your feedback and contribution. I'm happy to help you here.

**Reply by duckhead** [reply]
I did eventually get Okta to work this morning. I got on the Okta server, and found the logs on there, and that told me that I was logging in, but wasn't authorized for the app. I fumbled around a bit and found where I could change that and it worked. With that, I was able to get past the spinner, and I can see the correct webpage (and the lovely lady next to me confirmed it looked correct.) She's got a wishlist of things she'd like to see, and I might try to pick up a couple of her low hanging fruit to work on. Besides, I don't know Kotlin, and figured it would be interesting to work with. I really hate the val/var distinction though...

**Reply by JohnKallies** [reply]
Nice.

So you know, WGU is kind of early in its management of open source software and repos, and we aren't currently set up for PR's from external forks. I'm sure we can do that, but there hasn't been a need so far. We've managed this to date by a "branching" model, where we add interested contributors to the WGU Open Source GitHub organization, and you push your changes / PR to feature branches in our repo. We initially felt like that would help us understand the community and changes early; time will tell how that pays off.

Since this is where we are now, please advise if you want an invite to the WGU Open Source org? You would need this to share a PR; invites expire after 1 week.

**Reply by duckhead** [reply]
Thank you. I know the history, at least seen through your user's eyes ;-) I do have one small PR I could do, which is to upgrade Maven to 3.8.8. If you'd like that, I'll submit it.

**Reply by duckhead** [reply]
Ha, broke it again. I downloaded a fresh osmt, and ran through the setup. I copied over api/osmt-dev-stack.env from my previous environment, but not the api/osmt-dev-stack.env. And it passed ./osmt_cli.sh -v. Even though the test/ file still had the default values in it. 

Turns out, you are only checking for xxxxxx and the test/ file has XXXXXX. I changed the grep -q to grep -qi to fix it, but I could change the case on the test/ file just as easily.

This does bring up a question though. Should I use the same OpenID in both files? Or, use separate ones? I can sell myself on either approach.

**Reply by JohnKallies** [reply]
Good catch, Jason. Thank you. The API testing is a late-breaking change, and hasn't had much external flexing yet.

Let's please fix the XXXXX case, and keep the grep test tight? It's up to you how to allocate the Okta secrets there. If you have CI jobs running all the time, maybe you'd want to distribute them. For a local dev env, I'm guessing 1 is enough. They may rate-limit by IP anyways, so it might not matter.

Got for it with the Maven change in the Dockerfile as well, that's housekeeping we need to do.

**Reply by JohnKallies** [reply]
You were asking about the Docker images before, `base` and `build`. Those are really only used in support of OSMT Spring app images that go to DockerHub. You _can_ build them locally, but I never saw where that gets you much.

**Reply by duckhead** [reply]
It's all good, I like tinkering with these sorts of things. I'm still working with a fresh install to make sure I get the overall flow documented.

**Reply by duckhead** [reply]
OK, next issue (this one is cute.) I'm having some issues with my ElasticSearch container dying on me. I also notice that it's eating a lot more memory than I expect it to, like around 16GB., which is a couple orders of magnitude larger than my MySQL container.  I suspect those are related, but I've not been able to prove it. docker logs provides no hints as to what is going on, but I did find this, which is curious:

{"type": "server", "timestamp": "2023-09-28T22:13:32,179Z", "level": "INFO", "component": "o.e.i.g.ConfigDatabases", "cluster.name": "docker-cluster", "node.name": "48ca258a04bf", "message": "initialized default databases [[GeoLite2-Country.mmdb, GeoLite2-City.mmdb, GeoLite2-ASN.mmdb]], config databases [[]] and watching [/usr/share/elasticsearch/config/ingest-geoip] for changes" }
{"type": "server", "timestamp": "2023-09-28T22:13:40,161Z", "level": "INFO", "component": "o.e.i.g.GeoIpDownloader", "cluster.name": "docker-cluster", "node.name": "48ca258a04bf", "message": "updating geoip databases", "cluster.uuid": "Cz5UgcbrQmS4AcevuPUHOA", "node.id": "Yq7aemVxT4SYPdXdMx3m4w"  }

It seems to be loading GeoIP data, which is not a thing I would think that OSMT would want or need.  

It looks to be a known issue, and I'm not comfortable enough in Docker to implement the fix though:

https://github.com/elastic/elasticsearch/issues/76586

**Reply by JohnKallies** [reply, marked as answer]
To confirm, OSMT should not have any concept of GeoIP data. Best guess is that ou are getting it for free from ES or Spring Data. I have seen this, and I don't really know why.

It doesn't surprise me that ES uses a lot more than MySQL. In OSMT, MySQL is the source of truth, but all reads come from ES, and we have indexed for case insensitivity in a few places. That said, I would think this would bubble up in IO more than in RAM consumption.

I use a Mac, and within Docker Desktop (the GUI for configuring the Docker service), I have allowed 12GB RAM.
